### ISSUE TITLE
COPY creates folders with very restricted permissions

### ISSUE THREAD

**Issue by EricHripko** [containers/buildah#1534]
This issue was initially wrongfully reported under #1516 . The problem is that Buildah creates the folders in path for `COPY` using `mode=700` unlike Docker that uses `mode=755`. This is a problem for two reasons: running as non-root and Docker compatibility:
- If you're running as non-root _inside the container_, you now need to chase down all `COPY` statements and fix permissions with a `chmod`
- If you're migrating off Docker to Buildah, your images may get broken in unexpected ways

**Steps to reproduce the issue:**
1. Create an empty file: `touch distutils.cfg`
2. Create a `Dockerfile` that performs a `COPY` into a nested non-existent path
```Dockerfile
FROM docker.io/ubuntu
COPY distutils.cfg /usr/lib/python3.7/distutils/distutils.cfg
#                           ^- 'python3.7' and 'distutils' folders have to be created
#                               to copy the file
```
3. Build an image: `buildah bud -t proof-buildah .`
4. Create a container from it: `buildah from proof-buildah`
5. Check permissions: `buildah run proof-buildah-working-container ls -lah /usr/lib | grep python`

**Describe the results you received:**
`python3.7` folder has `mode=700`. See the full CLI output below:
```
$ buildah bud -t proof-buildah .
STEP 1: FROM docker.io/ubuntu
STEP 2: COPY distutils.cfg /usr/lib/python3.7/distutils/distutils.cfg
STEP 3: COMMIT proof-buildah
Getting image source signatures
Copying blob 762d8e1a6054 skipped: already exists
Copying blob e45cfbc98a50 skipped: already exists
Copying blob d60e01b37e74 skipped: already exists
Copying blob b57c79f4a9f3 skipped: already exists
Copying blob c7a123878960 done
Copying config 5ee072a37f done
Writing manifest to image destination
Storing signatures
--> 5ee072a37f58ff245eb40a410039bc989e93d03eba6e43aee0ce1cf910a12a7f
$ buildah from proof-buildah
proof-buildah-working-container
$ buildah run proof-buildah-working-container ls -lah /usr/lib | grep python
drwx------ 3 root root 4.0K Apr 18 21:05 python3.7
$ docker run --rm proof-buildah ls -lah /usr/lib | grep python
drwx------  3 root root 4.0K Apr 18 21:05 python3.7
```

**Describe the results you expected:**
`python3.7` folder has `mode=755`.  See the Docker CLI output below (for comparison):
```
$ docker build -t proof-docker .
Sending build context to Docker daemon   2.56kB
Step 1/2 : FROM docker.io/ubuntu
 ---> 94e814e2efa8
Step 2/2 : COPY distutils.cfg /usr/lib/python3.7/distutils/distutils.cfg
 ---> c1cebcb50780
Successfully built c1cebcb50780
Successfully tagged proof-docker:latest
$ docker run --rm proof-docker ls -lah /usr/lib | grep python
drwxr-xr-x  3 root root 4.0K Apr 18 21:04 python3.7
```

**Output of `rpm -q buildah` or `apt list buildah`:**

```
Listing... Done
buildah/xenial,now 1.8-1~dev~ubuntu16.04.2~ppa5 amd64 [installed]
```

**Output of `buildah version`:**

```
Version:         1.8-dev
Go Version:      go1.10.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:
Git Commit:
Built:           Sat Apr 13 16:42:38 2019
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
N/A
```

**Output of `cat /etc/*release`:**

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.1 LTS"
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
```

**Output of `uname -a`:**

```
Linux scw-35eb3f 4.5.7-generic #1 SMP Tue Jul 12 09:41:14 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information

# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary read-write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# AdditionalImageStores is used to pass paths to additional read-only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers (currently overlay, zfs, vfs, btrfs)
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"
```

**Comment by EricHripko** [comment]
My guess based on reading the code is below. Does this look right?
- `Builder.Add` decides that absolute path always exists and [doesn't create folders](https://github.com/containers/buildah/blob/master/add.go#L113)
- `Builder.Add` calls `Builder.copyFileWithTar`, which in turn [calls](https://github.com/containers/buildah/blob/973bb88ef1861a5b782c722c471dd79b6732852c/util.go#L112) `chrootarchive.CopyFileWithTarAndChown` from `containers.storage`
- `chrootarchive.CopyFileWithTarAndChown` creates an `Archiver` and [returns](https://github.com/containers/storage/blob/25923caa130d265d7d1cabade19da04bfd26e3f8/pkg/archive/archive.go#L1395) `Archiver.CopyFileWithTar` function
- `Archiver.CopyFileWithTar` doesn't find the full path, so it [creates it](https://github.com/containers/storage/blob/25923caa130d265d7d1cabade19da04bfd26e3f8/pkg/archive/archive.go#L1152) with `mode=700`

Assuming my investigation is correct, who's wrong?
- Is it `Builder.Add` that doesn't create folders for absolute paths?
- Or is it `Archiver.CopyFileWithTar` that uses `mode=700`?
- Or is it both?

**Comment by EricHripko** [comment]
I've managed to resolve my issue by modifying `Builder.Add` like this:
```go
	if destination != "" && filepath.IsAbs(destination) {
		dir, _ := filepath.Split(destination)
		if dir != "" {
			if err = idtools.MkdirAllAndChownNew(filepath.Join(dest, dir), 0755, hostOwner); err != nil {
				return errors.Wrapf(err, "error creating directory %q", filepath.Join(dest, dir))
			}
		}
		dest = filepath.Join(dest, destination)
	} else {
```
It simply gets the directory of the `COPY` target and makes sure that it exists. In my tests this seems to have worked for both file and folder cases. Should I create PR for this?

**Comment by TomSweeneyRedHat** [comment]
@EricHripko if you have the time, I'd love to see a PR

**Comment by EricHripko** [comment]
Will do as soon as I can 🙂

**Comment by rhatdan** [comment]
@EricHripko Is there just one more PR to get this closed, or can we close it now?

**Comment by EricHripko** [comment]
This'll be resolved once #1615 is merged. Thanks for confirming :+1:

**Comment by rhatdan** [comment]
It is being merged.
